source/grid/tria.cc

```diff
@@ -11641,35 +11641,35 @@ Triangulation<dim, spacedim>::create_triangulation(
 
 
   /*
-      When the triangulation is a manifold (dim < spacedim), the normal field
-      provided from the map class depends on the order of the vertices.
-      It may happen that this normal field is discontinuous.
-      The following code takes care that this is not the case by setting the
-      cell direction flag on those cell that produce the wrong orientation.
-
-      To determine if 2 neighbours have the same or opposite orientation
-      we use a table of truth.
-      Its entries are indexes by the local indices of the common face.
-      For example if two elements share a face, and this face is
-      face 0 for element 0 and face 1 for element 1, then
-      table(0,1) will tell whether the orientation are the same (true) or
-      opposite (false).
-
-      Even though there may be a combinatorial/graph theory argument to get
-      this table in any dimension, I tested by hand all the different possible
-      cases in 1D and 2D to generate the table.
+      When the triangulation is a manifold (dim < spacedim) and made of
+      quadrilaterals, the normal field provided from the map class depends on
+      the order of the vertices. It may happen that this normal field is
+      discontinuous. The following code takes care that this is not the case by
+      setting the cell direction flag on those cell that produce the wrong
+      orientation.
+
+      To determine if 2 neighbours have the same or opposite orientation we use
+      a table of truth. Its entries are indexes by the local indices of the
+      common face. For example if two elements share a face, and this face is
+      face 0 for element 0 and face 1 for element 1, then table(0,1) will tell
+      whether the orientation are the same (true) or opposite (false).
+
+      Even though there may be a combinatorial/graph theory argument to get this
+      table in any dimension, I tested by hand all the different possible cases
+      in 1D and 2D to generate the table.
 
       Assuming that a surface respects the standard orientation for 2d meshes,
       the tables of truth are symmetric and their true values are the following
-      1D curves:  (0,1)
-      2D surface: (0,1),(0,2),(1,3),(2,3)
+
+      - 1D curves:  (0,1)
+      - 2D surface: (0,1),(0,2),(1,3),(2,3)
 
       We store this data using an n_faces x n_faces full matrix, which is
      actually much bigger than the minimal data required, but it makes the code
      more readable.
 
     */
-  if (dim < spacedim)
+  if (dim < spacedim && all_reference_cells_are_hyper_cube())
```

Comment:
Maybe you can skip this for triangulation generation but are the results obtained on such meshes correct?

Reply:
If we check somewhere else that the orientation is correct in the case of triangles, then we should not have problems here. CGAL generates consistently oriented meshes. We should definitely check that they are consistently oriented for tringle meshes. How is this done in two dimensions + simplices now? If what we do when we create the grid is enough in 2d+simplices, then it should be ok also in this case. I think the issues with quads is that we have a highly non-standard orientation in our triangulation class (i.e., we do not base our triangulation on the half-edge datastructure, but on lexycographical ordering, and consistent orientation of edges across neighbors). This gives issues when extracting boundary triangulation (for example), since some faces will have the right orientation, and some other won't. I believe this is the reason for this check to be here in the first place. 
 
This is not the case for triangles. Triangles follow the same orientation of most other libraries (anti-clockwise), and so does our tetra-hedron classes, so each face is oriented consistently w.r.t. to each cell, and you are guaranteed that even if you extract the bundary of a tet-only grid, you get a consistently oriented grid. I believe the check for triangles (and tetra) should be that, on each edge `cell->face(f)->vertex_index(0) == cell->neighbor(f)->face(cell->neighbor(f)->neighbor_of_neighbor(f))->vertex_index(1)` is true, but I have to double check this.

Reply:
> We should definitely check that they are consistently oriented for tringle meshes. How is this done in two dimensions + simplices now?

I assumed that the meshes coming from a CAD tool are correctly oriented. But unfortunately, this is not always given as the issue https://github.com/dealii/dealii/issues/13346 showed. It was fixed in https://github.com/dealii/dealii/pull/13579. For the normal case `dim==spacedim`, we have the checks:

https://github.com/dealii/dealii/blob/5cdf7bd7c3679f9b57f8a0fff0cd4aefab89a80f/source/grid/tria.cc#L2309-L2323

and

https://github.com/dealii/dealii/blob/5cdf7bd7c3679f9b57f8a0fff0cd4aefab89a80f/source/fe/mapping_fe.cc#L1233-L1240

that will definitely be triggered if the mesh is not correctly orientated. 

> I believe the check for triangles (and tetra) should be that, on each edge cell->face(f)->vertex_index(0) == cell->neighbor(f)->face(cell->neighbor(f)->neighbor_of_neighbor(f))->vertex_index(1) is true, but I have to double check this.

Do you want to open an issue? I guess one can come up with a minimal test with a hand full of cells?

Reply:
It is unclear to me what we want to do. For the codimension zero case, it is clear that if a mesh has negative determinant, it should not be allowed. For co-dimension one grids, however, the J in JxW is defined as sqrt(det(DF^T.DF)), and it is positive regardless of the orientation of the normal (i.e, the previous expression is the same as computing the norm of the cross product of the tangent vectors, or the columns of the Jacobian). The orientation of the mesh in the codimension one case does not influence the sign of the determinant but just the orientation of the normal to the cell. Do we want to enforce a consistent orientation, or should we allow discontinuous normals? A moebious strip is not orientable, but it is a perfectly valid co-dimension one mesh... It won't make sense to ask for a continuous normal in many cases. I think we should remove the check, and let the user be responsible for the orientation.

Reply:
(I know my previous comments partly contradicts what I said earlier, but I did not have the moebious strip example in mind when I wrota that 
> If we check somewhere else that the orientation is correct in the case of triangles, then we should not have problems here. CGAL generates consistently oriented meshes.

Reply:
I am not sure what the influence of discontinuous normals are so I cannot really tell when/where it is the best to deal with them. Maybe you could post a simple example?